packages/flutter/lib/src/material/text_field.dart

```diff
@@ -1193,7 +1231,26 @@ class _TextFieldState extends State<TextField> with RestorationMixin implements
       setState(() {
         _isHovering = hovering;
       });
+      _statesController.update(MaterialState.hovered, _isHovering);
+    }
+  }
+
+  // Material states controller.
+  MaterialStatesController? _internalStatesController;
+
+  void _handleStatesControllerChange() {
+    // Force a rebuild to resolve MaterialStateProperty properties.
+    setState(() { });
+  }
+
+  MaterialStatesController get _statesController => widget.statesController ?? _internalStatesController!;
+
+  void _initStatesController() {
+    if (widget.statesController == null) {
+      _internalStatesController = MaterialStatesController();
     }
+    _statesController.update(MaterialState.disabled, !_isEnabled);
```

Comment:
why does it only update `.disabled` but not the other properties?

Reply:
Here and `didUpdateWidget` seem like the only reasonable places to listen for changes in `_isEnabled`, while the other state properties like `focus` (listens to changes in `FocusNode`) and `hover` (listens to `MouseRegion` callbacks) actively listen for changes. I'll also initialize the error state here.

Reply:
Focus is edge-triggered so you have to set the correct initial value otherwise it won't reflect the correct state until the next time you get notified of a focus state change right? The TextField can be given a `FocusNode` that may or may not have the focus when `initState` is called.